src/v/config/tests/config_store_test.cc

```diff
@@ -437,3 +440,54 @@ SEASTAR_THREAD_TEST_CASE(property_bind) {
     BOOST_TEST(bind3() == "newvalue4");
     BOOST_TEST(watch_count == 6);
 }
+
+SEASTAR_THREAD_TEST_CASE(property_conversion_bind) {
+    auto cfg = test_config();
+    BOOST_TEST(cfg.boolean() == false);
+    auto binding = cfg.boolean.bind<ss::sstring>(
+      [](bool v) { return v ? "true" : "false"; });
+    BOOST_TEST(binding() == "false");
```

Comment:
this is really cool. a few questions:

1. is the converted-to value cached in the binding? if not, does this encourage runtime conversion costs that a normal watch+lambda would largely avoid because it forces the user to store the computed result somewhere else when the value changes?

2. does this encourage config::* stuff to be more deeply integrated into implementations rather than being pushed to the edges? im not sure one is significantly better than the other, but it's nice in principle to have components with less coupling.

3. Aesthetically does it make sense to have `config::conversion_binding<bytes>` or `config::conversion_binding<jpeg>` which are config things but without a configuration representation? Admittedly these are extreme / unrealistic examples, but just wanted to ask.

Reply:
> 1. is the converted-to value cached in the binding?

yes, just like property value is cached in `binding<>`

> 2. does this encourage config::* stuff to be more deeply integrated into implementations rather than being pushed to the edges? im not sure one is significantly better than the other, but it's nice in principle to have components with less coupling.
>
> 3. Aesthetically does it make sense to have config::conversion_binding<bytes> or config::conversion_binding<jpeg> which are config things but without a configuration representation? Admittedly these are extreme / unrealistic examples, but just wanted to ask.

These two questions are connected. The fact that the binding, as it is used in the end, cares about the original property type, is what makes config::* stuff integrated into implementations. I did consider separating the binding from the cached value, so that the architecture, would be like `property<T>` -> `conversion_binding<U,T>(convert_func)` w/o cached value -> `bound_value<U>`, so that implementations can be actually agnostic of what `T` was. The issue is that we want to deliver this change in the next v23.1.x so I've decided not to explore that path yet, even though I like it more.

Reply:
makes sense